### PR TITLE
Prepare for new style generation

### PR DESCRIPTION
**Pull request by jrmi**

### CODE REVIEW

File: terra_layer/migrations/0058_migrate_style.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+@transaction.atomic
+def forward(apps, schema_editor):
+    Layer = apps.get_model("terra_layer", "Layer")
+
+    for layer in Layer.objects.all():
+        if layer.main_style:
+            print(f"Nothing to do for {layer}. Already migrated ?")
+            continue
+        # print(layer.layer_style)
```
Comment (frodrigo) [inline terra_layer/migrations/0058_migrate_style.py:141]:
```suggestion
```

File: terra_layer/models.py [hunk 2]
```diff
@@ -160,6 +160,9 @@ class Layer(models.Model):
 
     layer_style = JSONField(default=dict)
     layer_style_wizard = JSONField(default=dict)
```
Comment (frodrigo) [inline terra_layer/models.py:162]: Unused fields ?

File: terra_layer/tests/test_commands/test_layer_dump.py [hunk 3]
```diff
@@ -89,18 +90,19 @@ def test_command_launch(self, mock_stdout):
...
         )
         call_command("layer_dump", pk=layer.pk)
         self.assertEqual(
-            json.loads(mock_stdout.getvalue())["custom_styles"],
+            json.loads(mock_stdout.getvalue())["extra_styles"],
             [
                 {
-                    "style": {},
+                    "style_config": {},
+                    "style": {},  # TODO remove me
```
Comment (frodrigo) [inline terra_layer/tests/test_commands/test_layer_dump.py:102]: TODO ?

File: terra_layer/models.py (hunk 2 above)
Comment (jrmi) [reply terra_layer/models.py:162]: Yes, but i keep them until the next delivery to keep data if the migration is broken for some reason

**Comment by frodrigo** [comment]
Does it allows to generate proportional anything (like proportional circle) ?
The computation is based on circle area.

File: terra_layer/tests/test_commands/test_layer_dump.py (hunk 3 above)
Comment (jrmi) [reply terra_layer/tests/test_commands/test_layer_dump.py:102]: Removed. Necessary until old fields are removed.

File: terra_layer/style.py [hunk 4]
```diff
@@ -482,290 +509,579 @@ def gen_legend_circle(
...
+
+        steps.append(["==", field_getter, name])
+        steps.append(value)
+
+    steps.append(default_value or default_no_value)
+
+    if default_value is not None:
+        return ["case", ["has", data_field], steps, default_value]
+    else:
+        return steps
```
Comment (frodrigo) [inline terra_layer/style.py:644]: Use `match` in place of `case`.
https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#match

**Comment by jrmi** [comment]
> Does it allows to generate proportional anything (like proportional circle) ?
> The computation is based on circle area.



> Does it allows to generate proportional anything (like proportional circle) ?
> The computation is based on circle area.

Yes there is two proportionnal computation now:
- Radius proportionnal with is the previous behaviour with circle area
- Generic size proportionnal (line_width for example) that behave linearly

**Comment by frodrigo** [comment]
> > Does it allows to generate proportional anything (like proportional circle) ?
> > The computation is based on circle area.
> 
> Yes there is two proportionnal computation now:
> 
>     * Radius proportionnal with is the previous behaviour with circle area
> 
>     * Generic size proportionnal (line_width for example) that behave linearly

Note, Mapbox GL have no `square` primitive.

File: terra_layer/style.py (hunk 4 above)
Comment (jrmi) [reply terra_layer/style.py:644]: right !

**Comment by jrmi** [comment]
> 
> Note, Mapbox GL have no `square` primitive.

I use the same code as before, the one you did. I think it was working.

File: terra_layer/migrations/0058_migrate_style.py (hunk 1 above)
Comment (jrmi) [reply terra_layer/migrations/0058_migrate_style.py]: Done

File: terra_layer/tests/test_commands/test_layer_dump.py (hunk 3 above)
Comment (jrmi) [reply terra_layer/tests/test_commands/test_layer_dump.py]: Done

File: terra_layer/style.py (hunk 4 above)
Comment (jrmi) [reply terra_layer/style.py]: Done

File: terra_layer/models.py [hunk 5]
```diff
@@ -191,16 +198,24 @@ class Meta:
         ordering = ("order", "name")
 
     def save(self, wizard_update=True, **kwargs):
-        if self.layer_style_wizard and wizard_update:
-            style, legend_addition = generate_style_from_wizard(
-                self, self.layer_style_wizard
+        if (
+            "type" in self.main_style
+            and self.main_style["type"] == "wizard"
```
[the remark this answers is not in the document]
Comment (jrmi) [reply terra_layer/models.py:203]: Indeed

**Review by CFarcy — APPROVED** [review]

**Review by frodrigo — APPROVED** [review]

**Pull request merged**